src/coreclr/System.Private.CoreLib/src/Internal/Runtime/Binder/AssemblyName.cs

```diff
@@ -0,0 +1,389 @@
+        public ushort usRevisionNumber;       // Revision Number.
+        public byte* szLocale;               // Locale.
+    }
+
+    internal enum CorAssemblyFlags
+    {
+        afPublicKey = 0x0001,     // The assembly ref holds the full (unhashed) public key.
+
+        afPA_None = 0x0000,     // Processor Architecture unspecified
+        afPA_MSIL = 0x0010,     // Processor Architecture: neutral (PE32)
+        afPA_x86 = 0x0020,     // Processor Architecture: x86 (PE32)
+        afPA_IA64 = 0x0030,     // Processor Architecture: Itanium (PE32+)
+        afPA_AMD64 = 0x0040,     // Processor Architecture: AMD X64 (PE32+)
+        afPA_ARM = 0x0050,     // Processor Architecture: ARM (PE32)
+        afPA_ARM64 = 0x0060,     // Processor Architecture: ARM64 (PE32+)
+        afPA_NoPlatform = 0x0070,      // applies to any platform but cannot run on any (e.g. reference assembly), should not have "specified" set
+        afPA_Specified = 0x0080,     // Propagate PA flags to AssemblyRef record
+        afPA_Mask = 0x0070,     // Bits describing the processor architecture
+        afPA_FullMask = 0x00F0,     // Bits describing the PA incl. Specified
+        afPA_Shift = 0x0004,     // NOT A FLAG, shift count in PA flags <--> index conversion
+
+        afEnableJITcompileTracking = 0x8000, // From "DebuggableAttribute".
+        afDisableJITcompileOptimizer = 0x4000, // From "DebuggableAttribute".
+        afDebuggableAttributeMask = 0xc000,
+
+        afRetargetable = 0x0100,     // The assembly can be retargeted (at runtime) to an
+                                     //  assembly from a different publisher.
+
+        afContentType_Default = 0x0000,
+        afContentType_WindowsRuntime = 0x0200,
+        afContentType_Mask = 0x0E00, // Bits describing ContentType
+    }
+
+    internal enum AssemblyNameIncludeFlags
+    {
+        INCLUDE_DEFAULT = 0x00,
+        INCLUDE_VERSION = 0x01,
+        INCLUDE_ARCHITECTURE = 0x02,
+        INCLUDE_RETARGETABLE = 0x04,
+        INCLUDE_CONTENT_TYPE = 0x08,
+        INCLUDE_PUBLIC_KEY_TOKEN = 0x10,
+        EXCLUDE_CULTURE = 0x20,
+        INCLUDE_ALL = INCLUDE_VERSION
+                    | INCLUDE_ARCHITECTURE
+                    | INCLUDE_RETARGETABLE
+                    | INCLUDE_CONTENT_TYPE
+                    | INCLUDE_PUBLIC_KEY_TOKEN,
+    }
+
+    internal sealed unsafe class AssemblyName : AssemblyIdentity, IEquatable<AssemblyName>
+    {
+        public bool IsDefinition;
+
+        public AssemblyName(IntPtr pPEImage)
+        {
+            IdentityFlags |=
+                AssemblyIdentityFlags.IDENTITY_FLAG_CULTURE | AssemblyIdentityFlags.IDENTITY_FLAG_PUBLIC_KEY_TOKEN_NULL;
+
+            int* dwPAFlags = stackalloc int[2];
+            using IMdInternalImport pIMetaDataAssemblyImport = BinderAcquireImport(pPEImage, dwPAFlags);
+
+            Architecture = AssemblyBinderCommon.TranslatePEToArchitectureType(dwPAFlags);
```

Comment:
The managed binder should not need to deal with image architectures. The unmanaged runtime should validate very early whether the image is good for current architecture (and reject it if it is not). By the time we get to the managed binder, we should know that the image architecture is good.

Reply:
I started with focusing the data structures, and did not look at the caller. Leaving the cleanup later should be fine, since cleaning managed code is easy.